### ISSUE TITLE
Table reflection on old oracle 10.2 (best_effort) errors due to missing check

### ISSUE THREAD

**Issue by zeehio** [sqlalchemy/sqlalchemy#11557]
### Describe the bug

On an old Oracle 10.2 database I was trying to use reflection using the code below and I got an error due to a missing check on the oracle dialect.

Since it's a best effort support, I hope I can submit and get the PR #11558 reviewed and merged for this issue.

In summary:

The oracle dialect has two properties to check if some compression features are supported:

https://github.com/sqlalchemy/sqlalchemy/blob/83b0690f88f46c0c8c04f8f24b5a95f426eba12b/lib/sqlalchemy/dialects/oracle/base.py#L1530-L1537

The checks work as expected.

However those checks are not used when building the query:

https://github.com/sqlalchemy/sqlalchemy/blob/83b0690f88f46c0c8c04f8f24b5a95f426eba12b/lib/sqlalchemy/dialects/oracle/base.py#L2048-L2062

So the query fails with the error "compressed_for" is not a column 

In my PR, I just add a check for that compression or return a null instead. This makes  my code work again.

Since it's a best effort support and a small PR I hope there is no need for a test case.

### SQLAlchemy Version in Use

83b0690 (current main)

### DBAPI (i.e. the database driver)

oracledb

### Database Vendor and Major Version

Oracle Database 10,2

### Python Version

3.11

### Operating system

Windows

### To Reproduce

```python
from sqlalchemy import Table, MetaData, create_engine

dsn = "localhost:1521/TEST"
username = "username"
password = "pass"
engine = create_engine(
    "oracle+oracledb://:@",
    connect_args={"user": username, "password": password, "dsn": dsn}
)
metadata = MetaData()
my_table = Table("my_table", metadata, autoload_with=engine)
```

### Error

```
--------------------------------------------------------------------------
DatabaseError                             Traceback (most recent call last)
File sqlalchemy\engine\base.py:1967 in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1966     if not evt_handled:
-> 1967         self.dialect.do_execute(
   1968             cursor, str_statement, effective_parameters, context
   1969         )
   1971 if self._has_events or self.engine._has_events:

File sqlalchemy\engine\default.py:924, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
    923 def do_execute(self, cursor, statement, parameters, context=None):
--> 924     cursor.execute(statement, parameters)

File oracledb\cursor.py:701, in Cursor.execute(self, statement, parameters, **keyword_parameters)
    700 impl = self._impl
--> 701 impl.execute(self)
    702 if impl.fetch_vars is not None:

File src\\oracledb\\impl/thick/cursor.pyx:306, in oracledb.thick_impl.ThickCursorImpl.execute()

File src\\oracledb\\impl/thick/utils.pyx:456, in oracledb.thick_impl._raise_from_odpi()

File src\\oracledb\\impl/thick/utils.pyx:446, in oracledb.thick_impl._raise_from_info()

DatabaseError: ORA-00904: "A_TABLES"."COMPRESS_FOR": invalid identifier

The above exception was the direct cause of the following exception:

DatabaseError                             Traceback (most recent call last)

File sqlalchemy\util\deprecations.py:281, in deprecated_params.<locals>.decorate.<locals>.warned(fn, *args, **kwargs)
    274     if m in kwargs:
    275         _warn_with_version(
    276             messages[m],
    277             versions[m],
    278             version_warnings[m],
    279             stacklevel=3,
    280         )
--> 281 return fn(*args, **kwargs)

File sqlalchemy\sql\schema.py:432, in Table.__new__(cls, *args, **kw)
    425 @util.deprecated_params(
    426     mustexist=(
    427         "1.4",
   (...)
    430 )
    431 def __new__(cls, *args: Any, **kw: Any) -> Any:
--> 432     return cls._new(*args, **kw)

File sqlalchemy\sql\schema.py:486, in Table._new(cls, *args, **kw)
    484     return table
    485 except Exception:
--> 486     with util.safe_reraise():
    487         metadata._remove_table(name, schema)

File sqlalchemy\util\langhelpers.py:146, in safe_reraise.__exit__(self, type_, value, traceback)
    144     assert exc_value is not None
    145     self._exc_info = None  # remove potential circular references
--> 146     raise exc_value.with_traceback(exc_tb)
    147 else:
    148     self._exc_info = None  # remove potential circular references

File sqlalchemy\sql\schema.py:482, in Table._new(cls, *args, **kw)
    480 metadata._add_table(name, schema, table)
    481 try:
--> 482     table.__init__(name, metadata, *args, _no_init=False, **kw)
    483     table.dispatch.after_parent_attach(table, metadata)
    484     return table

File sqlalchemy\sql\schema.py:862 in Table.__init__(self, name, metadata, schema, quote, quote_schema, autoload_with, autoload_replace, keep_existing, extend_existing, resolve_fks, include_columns, implicit_returning, comment, info, listeners, prefixes, _extend_on, _no_init, *args, **kw)
    858 # load column definitions from the database if 'autoload' is defined
    859 # we do it after the table is in the singleton dictionary to support
    860 # circular foreign keys
    861 if autoload_with is not None:
--> 862     self._autoload(
    863         metadata,
    864         autoload_with,
    865         include_columns,
    866         _extend_on=_extend_on,
    867         _reflect_info=_reflect_info,
    868         resolve_fks=resolve_fks,
    869     )
    871 # initialize all the column, etc. objects.  done after reflection to
    872 # allow user-overrides
    874 self._init_items(
    875     *args,
    876     allow_replacements=extend_existing
   (...)
    879     all_names={},
    880 )

File sqlalchemy\sql\schema.py:894, in Table._autoload(self, metadata, autoload_with, include_columns, exclude_columns, resolve_fks, _extend_on, _reflect_info)
    892 insp = inspection.inspect(autoload_with)
    893 with insp._inspection_context() as conn_insp:
--> 894     conn_insp.reflect_table(
    895         self,
    896         include_columns,
    897         exclude_columns,
    898         resolve_fks,
    899         _extend_on=_extend_on,
    900         _reflect_info=_reflect_info,
    901     )

File sqlalchemy\engine\reflection.py:1526, in Inspector.reflect_table(self, table, include_columns, exclude_columns, resolve_fks, _extend_on, _reflect_info)
   1524 table_key = (schema, table_name)
   1525 if _reflect_info is None or table_key not in _reflect_info.columns:
-> 1526     _reflect_info = self._get_reflection_info(
   1527         schema,
   1528         filter_names=[table_name],
   1529         kind=ObjectKind.ANY,
   1530         scope=ObjectScope.ANY,
   1531         _reflect_info=_reflect_info,
   1532         **table.dialect_kwargs,
   1533     )
   1534 if table_key in _reflect_info.unreflectable:
   1535     raise _reflect_info.unreflectable[table_key]

File sqlalchemy\engine\reflection.py:2019, in Inspector._get_reflection_info(self, schema, filter_names, available, _reflect_info, **kw)
   2002         res = {}
   2003     return res
   2005 info = _ReflectionInfo(
   2006     columns=run(
   2007         self.get_multi_columns, check_filter_names_from_meth=True
   2008     ),
   2009     pk_constraint=run(self.get_multi_pk_constraint),
   2010     foreign_keys=run(self.get_multi_foreign_keys),
   2011     indexes=run(self.get_multi_indexes),
   2012     unique_constraints=run(
   2013         self.get_multi_unique_constraints, optional=True
   2014     ),
   2015     table_comment=run(self.get_multi_table_comment, optional=True),
   2016     check_constraints=run(
   2017         self.get_multi_check_constraints, optional=True
   2018     ),
-> 2019     table_options=run(self.get_multi_table_options, optional=True),
   2020     unreflectable=unreflectable,
   2021 )
   2022 if _reflect_info:
   2023     _reflect_info.update(info)

File sqlalchemy\engine\reflection.py:1992, in Inspector._get_reflection_info.<locals>.run(meth, optional, check_filter_names_from_meth)
   1990 try:
   1991     if has_result:
-> 1992         res = meth(filter_names=_fn, **kw)
   1993         if check_filter_names_from_meth and not res:
   1994             # method returned no result data.
   1995             # skip any future call methods
   1996             has_result = False

File sqlalchemy\engine\reflection.py:730, in Inspector.get_multi_table_options(self, schema, filter_names, kind, scope, **kw)
    692 r"""Return a dictionary of options specified when the tables in the
    693 given schema were created.
    694 
   (...)
    727 .. seealso:: :meth:`Inspector.get_table_options`
    728 """
    729 with self._operation_context() as conn:
--> 730     res = self.dialect.get_multi_table_options(
    731         conn,
    732         schema=schema,
    733         filter_names=filter_names,
    734         kind=kind,
    735         scope=scope,
    736         info_cache=self.info_cache,
    737         **kw,
    738     )
    739     return dict(res)

File sqlalchemy\dialects\oracle\base.py:1801, in OracleDialect._handle_synonyms_decorator.<locals>.wrapper(self, *args, **kwargs)
   1799 @wraps(fn)
   1800 def wrapper(self, *args, **kwargs):
-> 1801     return self._handle_synonyms(fn, *args, **kwargs)

File sqlalchemy\dialects\oracle\base.py:1807, in OracleDialect._handle_synonyms(self, fn, connection, *args, **kwargs)
   1805 def _handle_synonyms(self, fn, connection, *args, **kwargs):
   1806     if not kwargs.get("oracle_resolve_synonyms", False):
-> 1807         return fn(self, connection, *args, **kwargs)
   1809     original_kw = kwargs.copy()
   1810     schema = kwargs.pop("schema", None)

File sqlalchemy\dialects\oracle\base.py:2128, in OracleDialect.get_multi_table_options(self, connection, schema, filter_names, scope, kind, dblink, **kw)
   2124 if ObjectKind.TABLE in kind or ObjectKind.MATERIALIZED_VIEW in kind:
   2125     query = self._table_options_query(
   2126         owner, scope, kind, has_filter_names, has_mat_views
   2127     )
-> 2128     result = self._execute_reflection(
   2129         connection, query, dblink, returns_long=False, params=params
   2130     )
   2132     for table, compression, compress_for in result:
   2133         if compression == "ENABLED":

File sqlalchemy\dialects\oracle\base.py:1586, in OracleDialect._execute_reflection(self, connection, query, dblink, returns_long, params)
   1581         bindparam.literal_execute = True
   1583     query = visitors.cloned_traverse(
   1584         query, {}, {"bindparam": visit_bindparam}
   1585     )
-> 1586 return connection.execute(
   1587     query, params, execution_options=execution_options
   1588 )

File sqlalchemy\engine\base.py:1418, in Connection.execute(self, statement, parameters, execution_options)
   1416     raise exc.ObjectNotExecutableError(statement) from err
   1417 else:
-> 1418     return meth(
   1419         self,
   1420         distilled_parameters,
   1421         execution_options or NO_OPTIONS,
   1422     )

File sqlalchemy\sql\elements.py:515, in ClauseElement._execute_on_connection(self, connection, distilled_params, execution_options)
    513     if TYPE_CHECKING:
    514         assert isinstance(self, Executable)
--> 515     return connection._execute_clauseelement(
    516         self, distilled_params, execution_options
    517     )
    518 else:
    519     raise exc.ObjectNotExecutableError(self)

File sqlalchemy\engine\base.py:1640, in Connection._execute_clauseelement(self, elem, distilled_parameters, execution_options)
   1628 compiled_cache: Optional[CompiledCacheType] = execution_options.get(
   1629     "compiled_cache", self.engine._compiled_cache
   1630 )
   1632 compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
   1633     dialect=dialect,
   1634     compiled_cache=compiled_cache,
   (...)
   1638     linting=self.dialect.compiler_linting | compiler.WARN_LINTING,
   1639 )
-> 1640 ret = self._execute_context(
   1641     dialect,
   1642     dialect.execution_ctx_cls._init_compiled,
   1643     compiled_sql,
   1644     distilled_parameters,
   1645     execution_options,
   1646     compiled_sql,
   1647     distilled_parameters,
   1648     elem,
   1649     extracted_params,
   1650     cache_hit=cache_hit,
   1651 )
   1652 if has_events:
   1653     self.dispatch.after_execute(
   1654         self,
   1655         elem,
   (...)
   1659         ret,
   1660     )

File sqlalchemy\engine\base.py:1846, in Connection._execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1844     return self._exec_insertmany_context(dialect, context)
   1845 else:
-> 1846     return self._exec_single_context(
   1847         dialect, context, statement, parameters
   1848     )

File sqlalchemy\engine\base.py:1986, in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1983     result = context._setup_result_proxy()
   1985 except BaseException as e:
-> 1986     self._handle_dbapi_exception(
   1987         e, str_statement, effective_parameters, cursor, context
   1988     )
   1990 return result

File sqlalchemy\engine\base.py:2353,in Connection._handle_dbapi_exception(self, e, statement, parameters, cursor, context, is_sub_exec)
   2351 elif should_wrap:
   2352     assert sqlalchemy_exception is not None
-> 2353     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
   2354 else:
   2355     assert exc_info[1] is not None

File sqlalchemy\engine\base.py:1967, in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1965                 break
   1966     if not evt_handled:
-> 1967         self.dialect.do_execute(
   1968             cursor, str_statement, effective_parameters, context
   1969         )
   1971 if self._has_events or self.engine._has_events:
   1972     self.dispatch.after_cursor_execute(
   1973         self,
   1974         cursor,
   (...)
   1978         context.executemany,
   1979     )

File sqlalchemy\engine\default.py:924, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
    923 def do_execute(self, cursor, statement, parameters, context=None):
--> 924     cursor.execute(statement, parameters)

File oracledb\cursor.py:701, in Cursor.execute(self, statement, parameters, **keyword_parameters)
    699 self._prepare_for_execute(statement, parameters, keyword_parameters)
    700 impl = self._impl
--> 701 impl.execute(self)
    702 if impl.fetch_vars is not None:
    703     return self

File src\\oracledb\\impl/thick/cursor.pyx:306, in oracledb.thick_impl.ThickCursorImpl.execute()

File src\\oracledb\\impl/thick/utils.pyx:456, in oracledb.thick_impl._raise_from_odpi()

File src\\oracledb\\impl/thick/utils.pyx:446, in oracledb.thick_impl._raise_from_info()

DatabaseError: (oracledb.exceptions.DatabaseError) ORA-00904: "A_TABLES"."COMPRESS_FOR": invalid identifier
[SQL: SELECT a_tables.table_name, a_tables.compression, a_tables.compress_for 
FROM all_tables a_tables 
WHERE a_tables.owner = :owner_1 AND a_tables.table_name IN (:filter_names_1)]
[parameters: {'owner_1': 'username', 'filter_names_1': 'my_table'}]
(Background on this error at: https://sqlalche.me/e/20/4xp6)
```

**Comment by zeehio** [comment]
Thanks for triaging!

Looking at git history I found this bug is a regression introduced at: db08a699489c9b0259579d7ff7fd6bf3496ca3a2

Before that commit you can see the compression properties being used as expected.

**Comment by CaselIT** [comment]
Hi, 

Thanks for reporting. Yes I assumed that change was where the regression happened.

Other than this issue the rest of the reflection works on your oracle 10?

**Comment by zeehio** [comment]
Everything else I have used so far seems to work fine :-)

**Comment by sqla-tester** [comment]
**Sergio Oller Moreno** has proposed a fix for this issue in the **main** branch:

**Check if oracle compression-related features are supported before using them** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5408

**Comment by sqla-tester** [comment]
**Sergio Oller Moreno** has proposed a fix for this issue in the **rel_2_0** branch:

**Fix table reflection on oracle <10.2** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5409
